ori/volume/server/volume_create.go

```diff
@@ -17,260 +17,167 @@ package server
-			},
-			VolumeMode:       func(m corev1.PersistentVolumeMode) *corev1.PersistentVolumeMode { return &m }(corev1.PersistentVolumeBlock),
-			StorageClassName: &s.rookPoolStorageClass,
-		},
+	if oriVolume == nil {
+		return fmt.Errorf("volume is nil")
 	}
 
-	var imagePvc *corev1.PersistentVolumeClaim
-	if volume.Spec.Image != "" {
-		imageName := GetSanitizedImageName(volume.Spec.Image)
-		apiutils.SetImageAnnotation(pvc, volume.Spec.Image)
-
-		pvc.Spec.DataSourceRef = &corev1.TypedObjectReference{
-			APIGroup: pointer.String("snapshot.storage.k8s.io"),
-			Kind:     "VolumeSnapshot",
-			Name:     imageName,
-		}
-
-		size, err := getImageSize(ctx, volume.Spec.Image)
-		if err != nil {
-			return nil, fmt.Errorf("unable to get image size: %w", err)
-		}
-
-		imagePvc = &corev1.PersistentVolumeClaim{
-			TypeMeta: metav1.TypeMeta{
-				Kind:       "PersistentVolumeClaim",
-				APIVersion: "v1",
-			},
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      imageName,
-				Namespace: s.namespace,
-			},
-			Spec: corev1.PersistentVolumeClaimSpec{
-				AccessModes: []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
-				Resources: corev1.ResourceRequirements{
-					Requests: corev1.ResourceList{corev1.ResourceStorage: *size},
-				},
-				VolumeMode:       func(m corev1.PersistentVolumeMode) *corev1.PersistentVolumeMode { return &m }(corev1.PersistentVolumeBlock),
-				StorageClassName: pointer.String(s.rookPoolStorageClass),
-				//set DataSourceRef that populator picks up the pvc
-				DataSourceRef: &corev1.TypedObjectReference{
-					APIGroup: pointer.String(storagev1alpha1.SchemeGroupVersion.String()),
-					Kind:     "Volume",
-					Name:     pvc.Name,
-				},
-			},
-		}
-		metautils.SetLabel(imagePvc, volumev1alpha1.ManagerLabel, volumev1alpha1.VolumeCephletManager)
+	if err := s.validateVolumeClass(oriVolume); err != nil {
+		return fmt.Errorf("invalid volume class: %w", err)
+	}
+	log.V(2).Info("Validated class")
 
+	validatedImage, err := s.validateImage(ctx, log, oriVolume)
+	if err != nil {
+		return fmt.Errorf("invalid volume image: %w", err)
 	}
+	log.V(2).Info("Validated image")
 
-	if err := apiutils.SetObjectMetadata(pvc, volume.Metadata); err != nil {
-		return nil, err
+	if err := s.validateSize(oriVolume, validatedImage); err != nil {
+		return fmt.Errorf("invalid volume size: %w", err)
 	}
-	metautils.SetLabel(pvc, volumev1alpha1.ManagerLabel, volumev1alpha1.VolumeCephletManager)
-	apiutils.SetVolumeClassLabel(pvc, volume.Spec.Class)
+	log.V(2).Info("Validated size")
+
+	return nil
 
-	return &AggregateCephVolume{
-		Pvc:         pvc,
-		ImagePvc:    imagePvc,
-		VolumeClass: volumeClass,
-	}, nil
 }
 
-func (s *Server) handleImagePopulation(ctx context.Context, log logr.Logger, volume *AggregateCephVolume) error {
-	if volume.ImagePvc == nil {
-		return nil
-	}
+func (s *Server) prepareOSImage(ctx context.Context, log logr.Logger, osImageName string) (osImageId string, retErr error) {
```

Comment:
I'd recommend moving the functionality of downloading / preparing base images to another component. That component then can, given an image id, do the singleflight downloading.